lua/entities/gmod_wire_expression2/core/player.lua

```diff
@@ -676,3 +676,119 @@ end
 e2function entity lastDisconnectedPlayer()
 	return lastLeft
 end
+
+----- Deaths+Spawns, Dev: Vurv, 12/7/19 -----
+local DeathAlert = {} -- table of e2s that have runOnDeath(1)
+local RespawnAlert = {}
+local DeathList = { last = {NULL, NULL, NULL, 0} }
+local RespawnList = { last = {NULL,0} }
+
+hook.Add("PlayerDeath","Exp2PlayerDetDead",function(victim,inflictor,attacker)
+	local entry = {n={},ntypes={},s={},stypes={},size=0} -- default table
+	entry.s["Victim"]=victim
+	entry.s["Inflictor"]=inflictor
+	entry.s["Attacker"]=attacker
+	entry.s["Timestamp"]=CurTime()
+	entry.stypes["Victim"]="e"
+	entry.stypes["Inflictor"]="e"
+	entry.stypes["Attacker"]="e"
+	entry.stypes["Timestamp"]="n"
+	entry.size=4
+	DeathList[victim:EntIndex()] = entry -- victim's death is saved in victims death list.
+	DeathList.last = entry -- the most recent death's table is stored here for later use.
+    for ex,_ in pairs(DeathAlert) do -- loops over all chips in deathalert, ignores key.
+        if IsValid(ex) then
+            ex.context.data.runByDeath = entry
+            ex:Execute()
+            ex.context.data.runByDeath = nil
+        end
+    end
+end)
+
+hook.Add("PlayerSpawn","Exp2PlayerDetRespn",function(ply,transition)
+	local entry = {n={},ntypes={},s={},stypes={},size=0} -- default table
+	entry.s["Player"]=ply
+	entry.s["Timestamp"]=CurTime()
+	entry.stypes["Player"]="e"
+	entry.stypes["Timestamp"]="n"
+	entry.size=2
+	RespawnList[ply:EntIndex()] = entry
+	RespawnList.last = entry
+	for ex,_ in pairs(RespawnAlert) do
+		if IsValid(ex) then
+			ex.context.data.runByRespawned = entry
+			ex:Execute()
+			ex.context.data.runByRespawned = nil
+		end
+	end
+end)
+
+__e2setcost(5)
+
+--- If active is 0, the chip will no longer run on death.
+e2function void runOnDeath(number activate)
+	if activate ~= 0 then
+		DeathAlert[self.entity] = true
+	else
+		DeathAlert[self.entity] = nil
+	end
+end
+
+e2function void runOnSpawn(number activate)
+	if activate ~= 0 then
+		RespawnAlert[self.entity] = true
+	else
+		RespawnAlert[self.entity] = nil
+	end
+end
+
+--If ran by death, (defined in e2 data), gives 1 or 0 (ternary)
+e2function number deathClk()
+	return self.data.runByDeath and 1 or 0
+end
+
+e2function number spawnClk()
+	return self.data.runByRespawned and 1 or 0
+end
+
+e2function number lastDeathTime() -- returns when the last death happened
+	local lastD = DeathList.last
+	if not lastD then return {} end
+	return lastD.s["Timestamp"]
+end
+
+e2function number lastSpawnTime()
+	local lastS = RespawnList.last
+	if not lastS then return {} end
+	return lastS.s["Timestamp"]
+end
+
+e2function table lastDeath(entity ply)
+	if not IsValid(ply) then return {} end
```

Comment:
This in the rest of the functions returning `{}` as a default table will fail cause it's missing the table structure fields.